src/sources/pulsar.rs

```diff
@@ -0,0 +1,624 @@
+    .await;
+}
+
+/// Send the event stream created by the framed read to the `out` stream.
+async fn finalize_event_stream(
+    consumer: &mut Consumer<String, TokioExecutor>,
+    finalizer: &Option<OrderedFinalizer<FinalizerEntry>>,
+    out: &mut SourceSender,
+    mut stream: std::pin::Pin<Box<dyn futures_util::Stream<Item = Event> + Send + '_>>,
+    topic: String,
+    message_id: MessageIdData,
+    pulsar_error_events: &Registered<PulsarErrorEvent>,
+) {
+    match finalizer {
+        Some(finalizer) => {
+            let (batch, receiver) = BatchNotifier::new_with_receiver();
+            let mut stream = stream.map(|event| event.with_batch_notifier(&batch));
+
+            match out.send_event_stream(&mut stream).await {
+                Err(_error) => {
+                    emit!(StreamClosedError { count: 1 });
+                }
+                Ok(_) => {
+                    finalizer.add(FinalizerEntry { topic, message_id }, receiver);
+                }
+            }
+        }
+        None => match out.send_event_stream(&mut stream).await {
+            Err(_error) => {
+                emit!(StreamClosedError { count: 1 });
+            }
+            Ok(_) => {
+                if let Err(error) = consumer.ack_with_id(topic.as_str(), message_id).await {
+                    pulsar_error_events.emit(PulsarErrorEventData {
+                        msg: error.to_string(),
+                        error_type: PulsarErrorEventType::Ack,
+                    });
+                }
+            }
+        },
+    }
+}
+
+async fn handle_ack(
+    consumer: &mut Consumer<String, TokioExecutor>,
+    status: BatchStatus,
+    entry: FinalizerEntry,
+    pulsar_error_events: &Registered<PulsarErrorEvent>,
+) {
+    match status {
+        BatchStatus::Delivered => {
+            if let Err(error) = consumer
+                .ack_with_id(entry.topic.as_str(), entry.message_id)
+                .await
+            {
+                pulsar_error_events.emit(PulsarErrorEventData {
+                    msg: error.to_string(),
+                    error_type: PulsarErrorEventType::Ack,
+                });
+            }
+        }
+        BatchStatus::Errored => {
+            if let Err(error) = consumer
+                .nack_with_id(entry.topic.as_str(), entry.message_id)
+                .await
+            {
+                pulsar_error_events.emit(PulsarErrorEventData {
+                    msg: error.to_string(),
+                    error_type: PulsarErrorEventType::NAck,
+                });
+            }
+        }
+        BatchStatus::Rejected => {
+            if let Err(error) = consumer
+                .nack_with_id(entry.topic.as_str(), entry.message_id)
+                .await
+            {
+                pulsar_error_events.emit(PulsarErrorEventData {
+                    msg: error.to_string(),
+                    error_type: PulsarErrorEventType::NAck,
+                });
+            }
+        }
+    }
```

Comment:
We could collapse these two branches.
```suggestion
        BatchStatus::Errored | BatchStatus::Rejected=> {
            if let Err(error) = consumer
                .nack_with_id(entry.topic.as_str(), entry.message_id)
                .await
            {
                pulsar_error_events.emit(PulsarErrorEventData {
                    msg: error.to_string(),
                    error_type: PulsarErrorEventType::NAck,
                });
            }
        }
    }
```